SPIKE: toward correct-by-construction Zoe2 escrow

dckc opened the pull request:
## Description

@erights expressed an interest in a correct-by-construction methodology for writing the escrow service in Zoe 2. This spike explores various approaches.

### Background

The essential properties of the Zoe escrow service are the same as the 42 line escrow contract in the 2013 Dr. SES paper. The mint / purse pattern and amount math properties are essentially unchanged since the 2000 paper on capability-based financial instruments. 2015 marked the first steps toward formal reasoning about these patterns. In 2020, a _Chainmail_ formalism was developed, and in 2022 the techniques were applied more broadly and formalized in coq.

- “Necessity Specifications for Robustness.” 2022. Mackay, Julian, Susan Eisenbach, James Noble, and Sophia Drossopoulou. Proceedings of the ACM on Programming Languages 6 (OOPSLA2): 154:811-154:840. https://doi.org/10.1145/3563317.
- “Holistic Specifications for Robust Programs.” 2020. Drossopoulou, Sophia, James Noble, Julian Mackay, and Susan Eisenbach. In Fundamental Approaches to Software Engineering, edited by Heike Wehrheim and Jordi Cabot, 12076:420–40. Lecture Notes in Computer Science. Cham: Springer International Publishing. https://doi.org/10.1007/978-3-030-45234-6_21.
- “Reasoning about Risk and Trust in an Open World.” 2015. Drossopoulou, Sophia, James Noble, Toby Murray, and Mark S. Miller. Victoria University of Wellington. http://ecs.victoria.ac.nz/foswiki/pub/Main/TechnicalReportSeries/ECSTR15-08.pdf.
  - 2018-10 talk: [Towards Reasoning about Risk and Trust in an Open World \- YouTube](https://www.youtube.com/watch?v=lxf7HTxWluc&list=PLKr-mvz8uvUgybLg53lgXSeLOp4BiwvB2&index=27)
- “Swapsies on the Internet: First Steps towards Reasoning about Risk and Trust in an Open World.” 2015. Drossopoulou, Sophia, James Noble, and Mark S. Miller. In Proceedings of the 10th ACM Workshop on Programming Languages and Analysis for Security, 2–15. Prague Czech Republic: ACM. https://doi.org/10.1145/2786558.2786564.
- “The Digital Path: Smart Contracts and the Third World.” 2013. Miller, Mark S., and Marc Stiegler. In Markets, Information and Communication: Austrian Perspectives on the Internet Economy. http://papers.agoric.com/papers/the-digital-path-smart-contracts-and-the-third-world/abstract/.
- “Distributed Electronic Rights in JavaScript.” 2013. Miller, Mark S., Tom Van Cutsem, and Bill Tulloh. In Programming Languages and Systems, 7792:1–20. Lecture Notes in Computer Science. Berlin, Heidelberg: Springer. https://doi.org/10.1007/978-3-642-37036-6_1.
- “Capability-Based Financial Instruments.” 2001. Miller, Mark S., Chip Morningstar, and Bill Frantz. In , edited by Yair Frankel, 1962:349–78. Berlin, Heidelberg: Springer Berlin Heidelberg. https://doi.org/10.1007/3-540-45472-1_24.

_This bibliography is taken from the [escrow](https://www.zotero.org/connolly/tags/escrow/library)  tag in [my zotero library](https://www.zotero.org/connolly/library)_.

### Security / Documentation / Testing Considerations

Some examples / tests are important to confirm our understanding, but the goal is to not just test that many cases work but to formally verify that all cases work.

An important documentation consideration is the extent to which some formally verified code and its proof actually increase our confidence, given the risk of errors in stating correctness criteria.

### Scaling / Upgrade Considerations

not considered in this spike.

dckc committed:
SPIKE: ocap hello world (makeCounter) in dafny

dckc committed:
SPIKE: rights amplification in dafny: makeBrandPair

dckc committed:
SPIKE: passable in dafny

dckc committed:
SPIKE: amountMath in dafny

 - relations
 - exploring heap objects vs. ML-style values

dckc committed:
SPIKE: mint maker in dafny: facets

dckc committed:
SPIKE: heap stuff in dafny

dckc committed:
SPIKE(amountMath): algebraic properties only

never mind heap stuff

dckc committed:
chore: Escrow Exchange from 2013 Dr. SES paper

Fig. 2. The Escrow Exchange Contract

p 15

 - Miller, Mark S., Tom Van Cutsem, and Bill
   Tulloh. 2013. “Distributed Electronic Rights in JavaScript.” In
   Programming Languages and Systems, 7792:1–20. Lecture Notes in
   Computer Science. Berlin, Heidelberg:
   Springer. https://doi.org/10.1007/978-3-642-37036-6_1.

dckc committed:
chore(escrow2013): migrate eventual send from ! to E()

dckc committed:
chore(escrow2013): use standard Promise for Q

dckc committed:
chore(escrow2013): modern Agoric JS style

 - var -> let/const
 - linebreaks in ()

dckc committed:
factor out outcome.dfy

dckc committed:
WIP: JS.dfy: exploring Promises etc.

dckc committed:
chore: harden() escrow to conform to jessica Jessie parser

dckc committed:
chore(zoe): ignore spec/*.dll

dckc committed:
chore(zoe): elib spec verifies

dckc committed:
fixup! chore(zoe): ignore spec/*.dll

dckc committed:
fix(outcome.dfy): avoid allocating in Outcome

dckc committed:
refactor: factor Outcome out of makeIssuerKit

dckc committed:
style: lint passableObj.dfy

dckc committed:
fixup! style: lint passableObj.dfy

dckc committed:
WIP(elib): struggling with failure-compatible types

dckc committed:
fixup! style: lint passableObj.dfy

dckc committed:
docs(amountMath.dfy): brand objects vs. pure Amounts. hm.

dckc committed:
WIP: thoughts on dafny/JS FFI

dckc committed:
WIP: noodling on Promise<T> and Vat in dafny

dckc committed:
WIP: escrow basics using a port of ELib concepts to dafny

dckc committed:
WIP: escrowAgent transfer verifies with PromiseT<T> = T

dckc committed:
SPIKE: escrow2013.js Jessie AST

node ~/projects/jessica/lang/nodejs/jessparse.js spec/escrow2013.js |jq -c >spec/escrow2013.jessie.json

dckc commented:
## dafny tooling: ocaps hello world

The vs-code extension works pretty great, with 2 issues:
  1. installing the dotnet runtime is a bit tedious (see my [dafny bookmarks](https://www.diigo.com/user/dckc-madmode?query=dafny))
  2. the verifyer/solver unexpectedly start computing madly. ctrl-shift-p "Resart Dafny Server" works well sometimes.

I got `makeCounter` to verify.

 - 2023-07-26 10:11 675518097 SPIKE: ocap hello world (makeCounter) in dafny

I sort of got it compiled to javascript, but `Error: Cannot find module 'bignumber.js'` suggests I haven't installed something that the JS backend expects.

## ADTs: amountMath, passable

ADTs that don't use reference types (`object`, `class`, `trait`, ...) work like in ML/haskell:

 - 2023-07-26 10:12 477888a0f SPIKE: amountMath in dafny
 - 2023-07-26 10:12 09fe73558 SPIKE: passable in dafny

Verifying AmountMath properties seems pretty straightforward. Brands represent a potential complication. I think separating the amountMathHelpers layer might be the way to go.

## facets, heap access: mint maker

Dafny requires `modifies ...` specifications for methods that modify heap objects. I'm starting to get the hang of it, slowly.

 - 2023-07-26 10:14 69d596417 SPIKE: mint maker in dafny: facets

## Rights Amplification: unsolved

I haven't figured out rights amplification. I don't see weakmaps. I suspect I could get `makeBrandPair` working eventually, but my dafny-fu wasn't sufficient when I first tried:

 - 2023-07-26 10:11 8ba5d13fd SPIKE: rights amplification in dafny: makeBrandPair

dckc commented:
## Dr. SES escrow in modern Jessie

I went back to the 2013 formulation and made minimal edits to fit it to Jessie syntax:

 - 2023-07-29 15:02 236b6141e chore(escrow2013): modern Agoric JS style
 - 2023-07-29 15:00 0b02304e7 chore(escrow2013): use standard Promise for Q
 - 2023-07-29 14:59 d437a06a5 chore(escrow2013): migrate eventual send from ! to E()
 - 2023-07-29 14:47 88caae8cb chore: Escrow Exchange from 2013 Dr. SES paper

## Translate from Jessie?

The jessica toolkit suggests a possibility of writing in Jessie and analyzing the AST in something like dafny or ACL2.

 - 2023-08-13 18:09 d219ade90 SPIKE: escrow2013.js Jessie AST

## Jessie interpreter?

Another possibility is a Jessie interpreter, though verifying how code runs in an interpreter seems challenging.

cc @michaelfig @gibson042

dckc commented:
## Eventual send, Promises, async: notation only for now

How to work with promises in dafny? For example:

```js
   Q(decisionP).then( // setup phase 2
    _ => { E(dstPurseP).deposit(amount, escrowPurseP); },
    _ => { E(srcPurseP).deposit(amount, escrowPurseP); });
```

Methods are not first class in dafny. Dafny has lambdas, but they're for pure functions only. No method calls nor object allocation allowed. We can use techniques like from the early java days: a `Thunk` trait with an `apply()` method.
It's tedious, though: a separate class for each js arrow function.

There are hints of async support in dafny coming, but it's not there now. They have iterators, but I couldn't figure out how to do anything like promises with them.

Also, the Dr. SES escrow contract is full of eventual sends such as `E(dstPurseP).deposit(amount, escrowPurseP)`.

I tried java-esque techniques again: I ported enough of ELib to do `E.sendAll(dstPurseP, "deposit", [amount, escrowPurseP])`.

 - 2023-08-12 22:48 9e556b649 WIP: escrow basics using a port of ELib concepts to dafny

But then `Purse` has to have a `callAll()` method that dispatches `"deposit"` to a real deposit method. I can't imagine teaching the dafny verifier to see through all that.

So my current approach is to support `E(rx)` notation, but just make `E` the identify function. Likewise `await(p)`. Dafny failure idioms seem to work OK to work with catch vs. throw.

 - 2023-08-13 17:55 eb3f550e5 WIP: escrowAgent transfer verifies with PromiseT<T> = T

```dfy
      // setup phase 2
      match await(decisionP) {
        case Success(_) => {
          var ignore := E(dstPurseP).deposit(amount, escrowPurseP);
        }
        case Failure(_) => {
          var ignore := E(srcPurseP).deposit(amount, escrowPurseP);
        }
      }
```

dckc commented:
## ACL2 might to do enough of first-class functions

The limitations around `.then()` and `E(o).m()` got me looking at ACL2 again, and I found they have [limited second-order functionality](https://www.cs.utexas.edu/users/moore/acl2/manuals/current/manual/index-seo.php/ACL2____INTRODUCTION-TO-APPLY_42):

> The unreachable goal of this work is to allow the ACL2 user to pass 'functions' as objects and to apply them. That goal is unreachable because ACL2 remains a first order system. However, we can identify a certain syntactic class of ordinary ACL2 objects, called the `[tame](https://www.cs.utexas.edu/users/moore/acl2/manuals/current/manual/index-seo.php/ACL2____TAME) functions' (which are in fact not functions but are merely symbols and list expressions) and we can allow names of functions with certain tameness properties to be passed around and used as functions.

ACL2 has a [fty](https://www.cs.utexas.edu/users/moore/acl2/v8-5/combined-manual/?topic=ACL2____FTY) library that supports ML style data types. When I worked on [js2rho/JessieProgram.scala](https://github.com/dckc/js2rho/blob/master/src/main/scala/js2rho/JessieProgram.scala), I found that a datatype for the Jessie ADT is entirely straightfoward. Porting that to ACL2 looks like fun. It might be over-kill, though...

[escrow2013.jessie.json](https://github.com/Agoric/agoric-sdk/blob/dc-escrow-formal/packages/zoe/spec/escrow2013.jessie.json) is nothing but `[]` lists and strings. Replace the `[]` by `()` and it's an s-expression that ACL2 will eat just fine. I expect a lisp macro for `E` would be pretty straightforward to turn `(E dstPurseP "deposit" amount escrowPurseP)` into queuing `(marshal (list dstPurse "deposit" amount escrowPurse))` and returning a promise for the result. I suppose I'll have to brush up on liveslots marshaling and `HandledPromise`.

dckc committed:
SPIKE: jessie parse -> lisp sexpr

dckc committed:
SPIKE: thoughts on Jessie in idomatic ACL2

dckc commented:
Lean compiles to JavaScript, but I doubt it compiles to the kind of JavaScript we want. I wonder if the JavaScript monad from the KU research would work better:

 - [Sunroof: A Monadic DSL for Generating JavaScript \| SpringerLink](https://link.springer.com/chapter/10.1007/978-3-319-04132-2_5)

While looking for that, I found...
 - [ghcjs/jsaddle: JavaScript interface that works with GHCJS or GHC](https://github.com/ghcjs/jsaddle)

dckc commented:
@erights writes:

> Right now,
> - we do not yet have an “after” operation.
> - “point-to-point” should mean “vat of sender to receiving object”.
> - we still currently may be counting on vat-to-vat fifo from zcf to zoe, but we would like not to.

Maybe we could formally verify the last one? a bit of a moon-shot, perhaps.

dckc commented:
The OCaml Platform has [Formal Verification support](https://ocaml.org/docs/platform-roadmap#w22-formal-verification) that looks about as expressive as dafny. Plus, there are popular promise libraries.

I found a book, [OCaml Programming: Correct \+ Efficient \+ Beautiful](https://cs3110.github.io/textbook/cover.html) that covers both:

 - [8\.6\. Promises](https://cs3110.github.io/textbook/chapters/ds/promises.html)
 - [6\.7\. Proving Correctness](https://cs3110.github.io/textbook/chapters/correctness/proving_correctness.html)

Plus, it has [interactive exercises](https://github.com/Agoric/documentation/issues/582#issuecomment-1950957691).

dckc commented:
prompted by...

 - [Eio 1\.0 Release: Introducing a new Effects\-Based I/O Library for OCaml \| Tarides](https://tarides.com/blog/2024-03-20-eio-1-0-release-introducing-a-new-effects-based-i-o-library-for-ocaml/)
 - [Lambda Capabilities \- Thomas Leonard's blog](https://roscidus.com/blog/blog/2023/04/26/lambda-capabilities/)

I tried porting `escrow.js` to Js_of_ocaml. I got... something that compiles.

Every boundary crossing is painfully explicit...

https://gist.github.com/dckc/df28e7815e06ae9367c2cb1ec65dd0ad

and the resulting js, with `(js_of_ocaml (flags --enable=effects))` is mangled by a CPS transform.

dckc commented:
Are there Actor system libraries for idris? Or for acl2?

Ooh!  I/we should do it as goblins actors first, and then port that to acl2.

cc @davexunit @cwebber

dckc committed:
broke ground on escrow in goblins

dckc commented:
broke ground on escrow in [goblins](https://spritely.institute/files/docs/guile-goblins/0.13.0/index.html):

[escrow2013.scm](https://gist.github.com/dckc/217625ec0fcc31cf579d33e00b6f178a#file-escrow2013-scm) gist

davexunit commented:
Wow this is really cool to see @dckc!!

dckc committed:
chore: get .scm escrow code to compile

dckc committed:
docs: notes on goblins devtools

dckc committed:
WIP: escrow in scm

dckc committed:
WIP: ERTP, weak store, ...

dckc committed:
ERTP tests coming together

dckc committed:
toward escrow testin

dckc committed:
fixup race

dckc committed:
WOOT! escrow2013.scm happy-path works!

dckc commented:
2024-07-27 22:50 a632a34b10 WOOT! escrow2013.scm happy-path works!

dckc committed:
fixup: arity

dckc committed:
test: amount math equivalence relation

dckc committed:
fixup! WIP: ERTP, weak store, ...

dckc committed:
test: partial-order

dckc committed:
fixup! test: partial-order

dckc committed:
test: AmountMath add properties: commutative, ...

dckc committed:
test: subtract-ok

dckc committed:
test: minmax-ok, completing amount properties (sans brands)

dckc committed:
fixup! test: minmax-ok, completing amount properties (sans brands)

dckc committed:
chore: musing on endo passStyleOf

dckc committed:
test(amount-properties): parameterize over element generator

dckc committed:
feat(endo): basic patterns for ERTP

dckc committed:
feat(ertp): handle Amount with Brand

dckc committed:
test(endo): test Amount properties with Brand

dckc committed:
test(alice): make-issuer-kit -> makeIssuerKit (WIP)

other updates broke this, I think

dckc commented:
Oct 4... revisited F*...
https://www.diigo.com/user/dckc-madmode?query=%23JavaScript+%23verification

[Fully abstract compilation to JavaScript \| Proceedings of the 40th annual ACM SIGPLAN\-SIGACT symposium on Principles of programming languages](https://dl.acm.org/doi/10.1145/2429069.2429114)

uses lambdajs semantics... does the lambdajs syntactic dialect include Jessie?

dckc commented:
@erights gave [a talk on ERTP, Zoe, and Ownables](https://www.youtube.com/watch?v=O8Bx_Abj9Qc&list=PLzDw4TTug5O1A-tkPJe4HVq0VBPcNOMHm). He told me about subsequent discussion with J. Noble, and an idea of emphasizing exos and durable state in formal modelling.

This suggests taking ~40 line escrow2013.js and doing it with exos.

That might make it more Java-like and hence reduce the friction with Dafny.

We estimate it shouldn't grow by more than ~3x.

dckc committed:
chore: EscrowExchange exoClass (UNTESTED)

dckc commented:
> This suggests taking ~40 line escrow2013.js and doing it with exos.
> We estimate it shouldn't grow by more than ~3x.

1st draft (2cc8149) is 78 lines, including imports and types.

There's at least 1 piece of state that's surely not durable, so there's work to do...

dckc committed:
chore: EscrowExchange.run() returns paid promise too

dckc committed:
test: EscrowExchange.run() works in 1 case

dckc commented on packages/zoe/test/z2spec/escrow-exo.test.js:
```diff
@@ -0,0 +1,75 @@
+      Money: issuers.Money.makeEmptyPurse(),
+      Stock: issuers.Stock.makeEmptyPurse(),
+    };
+    const purse = purses[giveKW];
+    purse.deposit(kit[giveKW].mint.mintPayment(giveAmt));
+    const payment = purse.withdraw(giveAmt);
```
deposit/withdraw gymnastics is left-over from flailing when I couldn't figure out why balances were all 0 after `.run()`
I realized `decisionP` didn't entail depositing payouts; just escrow.

dckc committed:
refactor: makePromiseKit() -> withResolvers()

dckc committed:
refactor: hoist transfer; no need to be a method

dckc committed:
refactor: factored out SeatKit

dckc committed:
refactor: use SeatKit throughout

dckc committed:
refactor: test: never mind DepositFacet; polish

dckc committed:
fixup! refactor: factored out SeatKit

dckc committed:
feat: ownable seat, resolver (lightly tested)

lightly tested using manual mutation testing:

$ git diff .
diff --git a/packages/zoe/test/z2spec/escrow-exo.test.js b/packages/zoe/test/z2spec/escrow-exo.test.js
index 0806866854..223927377a 100644
--- a/packages/zoe/test/z2spec/escrow-exo.test.js
+++ b/packages/zoe/test/z2spec/escrow-exo.test.js
@@ -59,6 +59,8 @@ test('escrowExchange: Alice and Bob do simpleExchange ', async t => {
       const { seat, resolver } = await ex1.makeEscrowSeatKit(proposal, pmts);
       t.log(name, 'got seat, resolver', seat);

+      await resolver.makeTransferInvitation();
+
       const participate = simpleExchange(resolver);

       const deposit = E.when(seat.getPayouts(), payouts =>

    ℹ REJECTED from ava test("escrowExchange: Alice and Bob do simpleExchange "): (Error#1)
    ℹ Error#1: OwnableSeatResolver_caretaker revoked
    ℹ     at makeError (file:///home/connolly/projects/ag-escrow-formal/node_modules/ses/src/error/assert.js:350:61)
          at fail (file:///home/connolly/projects/ag-escrow-formal/node_modules/ses/src/error/assert.js:482:20)
          at Fail (file:///home/connolly/projects/ag-escrow-formal/node_modules/ses/src/error/assert.js:492:39)
          at Object.readOnly (file:///home/connolly/projects/ag-escrow-formal/packages/base-zone/src/prepare-revocable.js:128:23)
          at In "readOnly" method of (OwnableSeatResolver_caretaker revocable) [as readOnly] (file:///home/connolly/projects/ag-escrow-formal/node_modules/@endo/exo/src/exo-tools.js:171:24)
          at simpleExchange (file:///home/connolly/projects/ag-escrow-formal/packages/zoe/test/z2spec/escrow-exo.test.js:19:35)
          at Object.go (file:///home/connolly/projects/ag-escrow-formal/packages/zoe/test/z2spec/escrow-exo.test.js:64:27)

erights commented on packages/zoe/src/z2spec/escrow-exo.js:
```diff
@@ -0,0 +1,311 @@
+
+          return harden({
+            proposal,
+            currentAllocation,
```
Whether the invitation customDetails should include the currentAllocation is interesting. I was surprised to see it here. Let's discuss.

erights commented on packages/zoe/src/z2spec/escrow-exo.js:
```diff
@@ -0,0 +1,311 @@
+          const paid = await payoutsKit.subscriber.getUpdateSince();
+          return paid.value[kw];
+        },
+        async getPayouts() {
```
`getPayout` and `getPayouts` are unfortunately not compat with exclusive transferability of the right to collect payouts, unless we do something weird to revoke these atomically on ownership transfer. Since an exited seat is not transferable (which makes sense since it represents no remaining rights), we could somehow ensure that all payout promises dispensed before a transfer (and therefore before exit) do not become connected to the payouts that appear at exit. But this would need more ownables mechanism. I'm not sure what. Interesting!

erights commented on packages/zoe/src/z2spec/escrow-exo.js:
```diff
@@ -0,0 +1,311 @@
+
+        /** @param {unknown} completion */
+        async exit(completion) {
+          assert.fail('TODO');
```
```suggestion
          this.state.exiting = true;
          assert.fail('TODO');
```

erights commented on packages/zoe/src/z2spec/escrow-exo.js:
```diff
@@ -0,0 +1,311 @@
+          assert.fail('TODO');
+        },
+        fail(reason) {
+          assert.fail('TODO');
```
```suggestion
          this.state.exiting = true;
          assert.fail('TODO');
```

erights commented on packages/zoe/src/z2spec/escrow-exo.js:
```diff
@@ -0,0 +1,311 @@
+        getCurrentAllocation() {
+          return this.state.currentAllocation;
+        },
+        getExitSubscriber() {
+          return this.state.exitKit.subscriber;
+        },
```
Even for an exit subscriber, we should wonder what happens to an old exit subscriber on ownership transfer. It provides only info, not authority, so it isn't too weird for a previous owner to observe a future transfer. But it is weird, and IMO a privacy expectation violation if it happens by default.

erights commented on packages/zoe/src/z2spec/escrow-exo.js:
```diff
@@ -0,0 +1,311 @@
+        },
+
+        /** @param {unknown} completion */
+        async exit(completion) {
+          assert.fail('TODO');
+        },
+        fail(reason) {
+          assert.fail('TODO');
+        },
```
Since these are only requests to exit, that may or may not exit depending on exit terms and the contract, and in any case will not cause a synchronous exit, we need to worry that it would still leave the seat in an exclusively transferable state, even though a seat-initialized exit may be in progress from (what would be) the previous owner.

erights replied:
Maybe this is why your state variable is `exiting` rather than `exited`, so the seat becomes non-transferable as soon as an exit is requested from the seat? In which case

```suggestion
        async exit(completion) {
          this.seat.exiting = true;
          assert.fail('TODO');
        },
        fail(reason) {
          this.seat.exiting = true;
          assert.fail('TODO');
        },
```

erights commented on packages/zoe/src/z2spec/escrow-exo.js:
```diff
@@ -0,0 +1,311 @@
+
+        getInvitationCustomDetails() {
+          const { exiting, proposal, currentAllocation } = this.state;
+          !exiting || Fail`already exited`;
```
```suggestion
          !exiting || Fail`already exiting`;
```
or not, depending on the resolution of the exit race issues above.

erights commented on packages/zoe/src/z2spec/escrow-exo.js:
```diff
@@ -0,0 +1,311 @@
+        },
+
+        // repeat accessors for compat
+        hasExited() {
```
Depending on the exit race issue resolution above, this method might be misnamed. But fixing it would be a compat issue. What compat goals make sense for zoe2 vs zoe1?

erights commented on packages/zoe/src/z2spec/escrow-exo.js:
```diff
@@ -0,0 +1,311 @@
+        },
+      },
+      /** The right to allocate payouts (subject to offer safety) */
+      resolver: {
+        readOnly() {
+          const { readOnly } = this.facets;
+          return readOnly;
+        },
```
Raises the same privacy expectation issue wrt previous owners finding out about exits after ownership transfer. But in this case, wrt resolver ownership transfer, which I don't think makes the issue either better or worse.

erights commented on packages/zoe/src/z2spec/escrow-exo.js:
```diff
@@ -0,0 +1,311 @@
+
+          this.state.exiting = true;
+          exitKit.publisher.finish(completion);
+          await deeplyFulfilledObject(payoutsKit.subscriber.getUpdateSince());
```
I'm not sure I understand this line.

erights commented on packages/zoe/src/z2spec/escrow-exo.js:
```diff
@@ -0,0 +1,311 @@
+          const { exiting, proposal, currentAllocation } = this.state;
+          !exiting || Fail`already exited`;
+
+          return harden({
+            proposal,
+            currentAllocation,
+          });
```
How does one distinguish an invitation for an ownable seat vs an ownable seat resolver?

erights commented on packages/zoe/src/z2spec/escrow-exo.js:
```diff
@@ -0,0 +1,311 @@
+    }),
+    {
+      /**
+       * TODO: use brands instead of keywords?
```
I sure hope we can do that!

erights commented on packages/zoe/src/z2spec/escrow-exo.js:
```diff
@@ -0,0 +1,311 @@
+       */
+      async addIssuer(keyword, issuer) {
+        !(keyword in this.state.issuers) ||
+          Fail`keyword alread used: ${keyword}`;
```
```suggestion
          Fail`keyword already used: ${keyword}`;
```

erights replied:
Redundant with test on line 246 below?

erights commented on packages/zoe/src/z2spec/escrow-exo.js:
```diff
@@ -0,0 +1,311 @@
+       * @param {Keyword} keyword
+       * @param {Issuer} issuer
+       */
+      async addIssuer(keyword, issuer) {
```
Please validate that `keyword` is a proper keyword

erights commented on packages/zoe/src/z2spec/escrow-exo.js:
```diff
@@ -0,0 +1,311 @@
+          Fail`keyword alread used: ${keyword}`;
+        const purse = await E(issuer).makeEmptyPurse();
+        const { issuers, escrowPurses } = this.state;
+        !(keyword in issuers) || Fail`keyword alread used: ${keyword}`;
```
```suggestion
        !(keyword in issuers) || Fail`keyword already used: ${keyword}`;
```

erights commented on packages/zoe/src/z2spec/escrow-exo.js:
```diff
@@ -0,0 +1,311 @@
+        return this.state.issuers;
+      },
+      /**
+       * TODO: with the escrow service separate from the contract host,
+       * is the exit rule relevant to the escrow service?
```
Good question!

erights commented on packages/zoe/src/z2spec/escrow-exo.js:
```diff
@@ -0,0 +1,311 @@
+        assertKeywordSubset('payment', payments, escrowPurses);
+        const giveKeys = keys(proposal.give);
+
+        // TODO: partial failure
```
This is the issue of a misbehavior issuer blocking assets from well behaved issuers, right?

https://github.com/Agoric/agoric-sdk/issues/1271

erights commented on packages/zoe/src/z2spec/escrow-exo.js:
```diff
@@ -0,0 +1,311 @@
+          );
+          // TODO: failure?
+          // TODO: expose payments publisher?
+          return E.when(payee.receive(payouts), () => {});
```
Why is this gated on an `exitSubscriber.subscribeAfter()` ?

dckc committed:
spec: add Iris opam helper makefile

dckc committed:
spec: add initial Coq smoke tests

dckc committed:
spec: add JSON Iris language skeleton

dckc committed:
spec: add Justin syntax and refinement scaffold

dckc committed:
spec: add Coq JSON concrete syntax parser

dckc committed:
docs: toward Jessie formalization in Iris (typo: idris)

dckc committed:
spec: add executable Justin and Jessie semantics

dckc committed:
spec: add tiny Coq Justin concrete parser

dckc committed:
spec: add Jessie regression bundle

dckc committed:
spec: finish Justin Iris language layer

dckc committed:
spec: add bigint and typeof null regressions

dckc committed:
spec: make freeze shallow and harden deep

dckc committed:
spec: get makeCounter working through assert

dckc committed:
spec: parse makeCounter surface syntax in Coq

dckc committed:
spec: layer Jessie surface syntax over Justin

dckc committed:
spec: add authority-split counter capabilities

dckc committed:
spec: move counter case study out of Justin core

dckc committed:
spec: move counter Iris facts out of Justin Iris core

dckc committed:
spec: move makeCounter execution into counter case study

dckc committed:
spec: share generic parser utilities

dckc commented:
@erights @michaelfig This evening I think I formalized a good chunk of Jessie in [Iris](https://iris-project.org/), "A Higher-Order Concurrent Separation Logic Framework, implemented and verified in the Rocq Prover".

Prompted by [a note on having fun with Idris](https://types.pl/@edwinb/116428517420906766), I revisited this spike, exploring Idris and then lean4 ([lean4-escrow-formal.md](https://gist.github.com/dckc/60f4e73f26784e3120983a9726d9731d#file-lean4-escrow-formal-md)) and then [Refined Rust](https://rust-formal-methods.github.io/meetings/refinedrust/).

Iris seemed like a good match for the formal properties we want (separation is good for "if you don't have the reference, you can't invoke it" and the escrow service has non-trivial async / concurrency properties). I was intimidated by Rocq / Coq, but... it's clearly the most mature basis for Iris, which means it's best for working with LLMs. So I got one LLM to write [the jessie-iris.md plan/prompt](https://github.com/Agoric/agoric-sdk/blob/dc-escrow-formal/packages/zoe/spec/jessie-iris.md), and lo, it seems to work. I think we have a formal verification of [the `makeCounter` separation of duties](https://docs.agoric.com/guides/js-programming/hardened-js#separation-of-duties) example.

dckc committed:
spec: refactor Jessie parser and counter layering

dckc committed:
spec: record forged-authority counterexamples

dckc committed:
spec: split source literals from runtime authority

dckc committed:
spec: add public Jessie client boundary

dckc committed:
spec: prove counter capability trace monotonicity

dckc committed:
spec: prove reachable counter authority polarity

dckc committed:
spec: add primitive connectivity model

dckc committed:
spec: prove builtins preserve reachable authority

dckc committed:
spec: add evaluator connectivity scaffolding

dckc committed:
spec: add closed-state and primitive reachability lemmas

dckc committed:
spec: strengthen evaluator connectivity scaffolding

dckc committed:
spec: note evaluator-step proof backtrack

dckc committed:
spec: split evaluator frame lemmas further

dckc committed:
spec: checkpoint public client theorem scaffolding

dckc committed:
spec: add real Jessie closure execution layer

dckc committed:
spec: factor builtin primitive environment into jessie_lib

dckc committed:
spec: document Iris work and improve Coq dev tooling

dckc committed:
spec: trim dev-tools note to point at iris.mk help

dckc committed:
spec: factor env updates and clarify makeCounter fixture naming

dckc committed:
spec: move shared surface runtime helpers into jessie_lib

dckc committed:
spec: move counter allocation out of generic Justin

dckc committed:
spec: remove counter-specific names from generic boundary examples

dckc committed:
spec: drop unused makeCounter closed-state lemma from step connectivity

dckc committed:
spec: split generic reachability from counter case study

dckc committed:
spec: attach connectivity contract to primitive handlers

dckc committed:
spec: move builtin conservativity proofs into jessie_lib

dckc committed:
spec: split JSON parsing out of generic parser helpers

dckc committed:
spec: separate JSON parsing from parser utilities

dckc committed:
spec: parse Justin field access and calls in core parser

dckc committed:
spec: document parser entry points against Jessica grammar goals

dckc committed:
spec: add get-result reachability helper

dckc committed:
spec: factor expression reachability through explicit roots

Introduce expr_root/exprs_root/fields_root as a syntactic root layer between core expressions and dynamic-authority reachability. Add the two bridge theorems in both directions so later preservation proofs can reason about old dynamic authority by first recovering a concrete root value, instead of inverting expr_reaches_dyn directly through CoreLetIn and other syntactic forms.

dckc committed:
spec: factor substitution reachability through expression roots

Extend the new expr_root layer with a substitution-source proof and the first useful corollary, expr_reaches_dyn_subst_old. The key engineering change is to handle arg/field lists through helper fixpoints parameterized by the expression case, rather than trying to force a single illegal mutual Fixpoint across core_expr and plain lists.

This is directly in line with the standard advice to insert an intermediate variable/root layer before tackling substitution. References: Software Foundations STLC properties chapter on why naive substitution inductions are too weak (<https://softwarefoundations.cis.upenn.edu/plf-current/StlcProp.html>, also <https://softwarefoundations.cis.upenn.edu/sf-4.0/StlcProp.html>), and the Iris issue on peeling off awkward proof layers before relying on unification (<https://gitlab.mpi-sws.org/iris/iris/-/issues/251>).

dckc committed:
spec: derive the let-value reachability bridge from root factoring

Build on the expr_root factoring to prove three concrete substitution lemmas:

- expr_root_subst_source
- expr_reaches_dyn_subst_sources
- let_value_step_reaches_source

This isolates the CoreLetIn/value-step case that was blocking the old-pid preservation proof. After this change, the substitution case no longer needs a brittle direct induction on expr_reaches_dyn; it is discharged by recovering a root, tracing that root back through subst, and then rebuilding reachability on the original let expression.

dckc commented:
> ... I think I formalized a good chunk of Jessie in Iris ...

Looks like I'm doing what Swasey and co did in 2017; it's number 8 among the [Iris pubs](https://iris-project.org/#publications).

 - Robust and Compositional Verification of Object Capability Patterns
David Swasey, Deepak Garg, Derek Dreyer
In OOPSLA 2017: ACM SIGPLAN Conference on Object-Oriented Programming, Systems, Languages, and Applications
Recipient of OOPSLA 2017 Distinguished Paper Award

I bookmarked it right when it came out, in fact. https://diigo.com/012e03k

dckc committed:
Reset spec to a minimal OCPL-style makeCounter core

Drop the old Jessie/Iris proof stack and replace it with a tiny language
containing just the constructs needed for makeCounter: fn/let, refs,
load/store, objects with last-wins fields, field get, numeric literals,
addition/subtraction, and assignment sugar.

Re-express makeCounter and cUp in that language, update the build inputs,
and trim the design note to the new minimal starting point.

dckc committed:
spec: checkpoint HeapLang pivot and restore process notes

Restore the design-note sections on dev tools and commit discipline, keeping the dev-tools section brief and pointing readers at iris.mk help for exact commands.

Extend the minimal Jessie syntax with assert and > so the goodness-bit client shape can be expressed in the local scaffold, and add the executable monitor / robust-safety scaffolding files to the default build.

Split generic OCPL-style attacker/goodness machinery from the counter-specific client instantiation, so jessie_robust stays generic while jessie_counter_robust carries the makeCounter-specific checked-client examples.

Add a direct HeapLang encoding of makeCounter using coq-iris-heap-lang, including checked-client and overshared-client programs, WP proofs for the hole and single-incr attacker cases, and matching adequacy corollaries. This establishes the main strong-reuse path on top of Iris/HeapLang instead of the bespoke evaluator as the proof foundation.

dckc committed:
spec: vendor curated OCPL reference subtree

Add an ocpl/ subtree containing the project-local reference material for Swasey, Garg, and Dreyer's 'Robust and Compositional Verification of Object Capability Patterns' (OOPSLA 2017).

Include a README with citation, upstream links, licensing notes, and the specific reasons this material is relevant to the Jessie counter proof. Also include a Jessie adaptation plan that maps OCPL notions such as low values, adversarial contexts, and good_state/is_good onto the intended Jessie theorem shape.

Vendor the proof-relevant upstream files that the current design work refers to: heap.v, robust_safety.v, adequacy.v, readonly.v, and the upstream README. This keeps the key OCPL theorem pattern available in-tree without depending on the raw full extraction under _research/.

dckc committed:
spec: add OCPL-shaped HeapLang adversary scaffold

Port the first reusable pieces of OCPL robust-safety structure into the
modern HeapLang-based Jessie spec tree.

- add jessie_ocpl_heaplang.v
- define OCPL-style adversarial values and expressions over modern HeapLang
- define a single-hole context grammar and context filling for the relevant
  HeapLang fragment
- record the modern HeapLang assertion mismatch explicitly: assert is a
  library value, so adversaries must exclude that value rather than a
  dedicated Assert AST node
- keep jessie_counter_heaplang.v as the concrete counter case-study file,
  while moving the reusable proof direction into the new generic scaffold
- update _CoqProject, iris.mk, and jessie-iris.md accordingly

This keeps the direct HeapLang path green while shifting the main proof plan
back toward OCPL-style reuse instead of the bespoke higher-order attacker
hypothesis.

dckc committed:
spec: add OCPL-style verified/safe layer scaffold

Continue the strong-reuse HeapLang path by extending the OCPL adaptation
file with the next semantic layer from robust_safety.v.

- record a TODO to add a tiny Jessie surface with Obj/Get that desugars to
  HeapLang terms
- add verified and safe predicates to jessie_ocpl_heaplang.v
- define them in the OCPL style: verified code is closed and satisfies a
  chosen postcondition; safe contexts preserve that postcondition when filled
  with verified code
- prove the base lemma safe_hole

This keeps the build green while moving the proof target away from bespoke
attacker hypotheses and toward the actual OCPL robust-safety structure.

dckc committed:
spec: extend OCPL HeapLang scaffold with context machinery

Continue the strong-reuse OCPL port in jessie_ocpl_heaplang.v.

- keep the Jessie-surface desugaring TODO in place, but stay on HeapLang for
  the proof foundation
- add the first semantic robust-safety layer: verified, safe, and safe_hole
- add a dedicated single-hole context grammar for the HeapLang fragment in use
- add executable conversions between that context grammar and selected
  HeapLang evaluation-context forms
- keep prophecy/Resolve support out of the round-trip proof path for now,
  since it is not on the makeCounter critical path and was causing proof noise

This preserves the green build while moving the project closer to the actual
OCPL robust-safety structure instead of bespoke attacker hypotheses.

dckc committed:
spec: move OCPL port work under ocpl/

Relocate the generic OCPL-to-modern-Iris adaptation work from the Jessie
case-study namespace into the ocpl/ subtree, where it belongs.

- add ocpl/modern_robust_safety.v as the current landing zone for the modern
  HeapLang robust-safety scaffold
- remove jessie_ocpl_heaplang.v
- update _CoqProject and iris.mk to build the port from ocpl/
- update jessie-iris.md and ocpl/README.md to point at the port in its new
  canonical location

This keeps the build green while aligning the file layout with the intended
strong-reuse workflow: OCPL code is ported in ocpl/, Jessie files instantiate
that port.

dckc committed:
spec: align OCPL port file with upstream interfaces

Continue the port inside ocpl/modern_robust_safety.v by making it look more
like the original OCPL robust_safety.v structure.

- add the Adversarial class and adv notation layer
- expose expr, val, and ctx instances through that interface
- rename the core predicates to OCPL-style names such as advexpr, advval,
  and advctx while preserving the current modern-Iris compatibility layer
- update class-field syntax to current Coq conventions

This keeps the port compiling on the current toolchain while reducing the gap
between the vendored upstream file and the local adapted copy.

dckc committed:
Port OCPL low-integrity layer to modern HeapLang

Add ocpl/modern_heap.v as the start of a modernized replacement for the upstream OCPL heap.v layer against the current Iris/HeapLang APIs.

Define the generic LowIntegrity class there, then instantiate it for current HeapLang values and environments using an Iris contractive fixpoint rather than the old on_val helper.

Add supporting low-environment lemmas used by the OCPL robust-safety development, including lookup, insert, delete, singleton, and empty cases.

Wire the ocpl subtree into the build with a dedicated logical path via -Q ocpl OCPL so vendored OCPL port files can depend on each other cleanly.

Refactor ocpl/modern_robust_safety.v to import the new low-integrity layer, use low locations instead of an ad hoc loc_ok parameter, and move the verified/safe definitions closer to upstream robust_safety.v by quantifying over low substitution environments through subst_map.

Also add the safe_alt lemma and keep the full spec build green under make -f iris.mk build.

dckc committed:
Port OCPL stuckness helpers to modern HeapLang

- add ocpl/modern_lifting.v with modernized shape predicates and
  MaybeStuck wp_stuck_* lemmas for variables, non-function apps,
  failing unary and binary operators, if, fst, snd, and case
- adapt the helper proofs to current HeapLang syntax, base_step rules,
  and current Iris lifting APIs
- record the one semantic mismatch explicitly by omitting the old
  wp_stuck_rec_open lemma, since modern HeapLang reduces Rec to RecV
  even for open bodies
- wire modern_lifting into _CoqProject and iris.mk so the default
  OCPL port build includes the new helper layer
- keep modern_robust_safety parameterized over stuckness instead of
  hard-coding NotStuck, matching the old OCPL progress-bit structure

dckc committed:
Port OCPL on_val layer to modern HeapLang

Add ocpl/modern_on_val.v as the modernized port of OCPL's value-
elimination layer and include it in the default spec build.

Modernize the low-value fixpoint in ocpl/modern_heap.v so recursive
values quantify over arbitrary masks and use MaybeStuck, then expose the
pair/injection elimination and introduction lemmas needed by the on_val
port. Add the minimal proofmode pair instances that the port relies on.

Wire modern_on_val into ocpl/modern_robust_safety.v and leave an explicit
TODO at the next reuse point: porting the confined constructor lemmas on
top of the new on_val layer.

Verification: make -f iris.mk build

dckc committed:
Port first modern OCPL confined lemmas and fix LowIntegrity instance use

Update ocpl/modern_robust_safety.v to follow the upstream OCPL confined setup more closely by deriving value and environment low-integrity instances from the canonical location-based OCPLModernHeap instances instead of quantifying over arbitrary LowIntegrity Σ val/env assumptions.

Add the first compiled confined lemmas on the modern stack: confined_var plus the application, pair, fst, snd, inl, and inr constructor lemmas. These now reuse the modernized wp_on_val_* bind lemmas directly under MaybeStuck, matching the upstream robust_safety.v proof shape more closely.

Extend ocpl/modern_on_val.v with raw wrapper lemmas for the value-level wp_on_val rules (app, pair, fst, snd, inl, inr, case). These wrappers keep the modern on_val layer usable from later confined ports and make the value-elimination interface explicit in the vendored ocpl/ subtree.

Leave confined_case as an explicit TODO in modern_robust_safety.v. The full build remains green with make -f iris.mk build.

dckc committed:
Add modern OCPL recursion and binder helper infrastructure

Extend ocpl/modern_heap.v with the missing recursion-introduction direction for low values (low_rec_intro_ent/raw), so the modern port can construct low recursive closures without unfolding low_val by hand in every proof.

Add low_env_binder_insert to ocpl/modern_heap.v. This packages the binder-aware environment extension step that the upstream confined_rec proof needs when inserting recursive closures and arguments under named or anonymous binders.

Register advval and advexpr as persistent in ocpl/modern_robust_safety.v using a mutual proof, aligning the modern port with upstream OCPL's adversarial syntax structure and making recursive proofs reusable under Iris proofmode.

The build remains green with make -f iris.mk build. The immediate next step is to finish porting confined_rec on top of these helpers.

dckc committed:
spec: refine Jessie/Iris slide deck around separation of duties

Recast jessie-iris.md as a slide deck and refine it for a technically strong
reader outside proof-assistant culture.

This revision:
- centers the deck on the separation-of-duties claim for makeCounter
- defines OCPL and Iris with in-context citations
- compresses the opening and separates prior work, proof foundation, and plan
- folds references into the relevant slides
- clarifies concurrency, mutable-state notation, and the modern-vs-2017 OCPL
  port mismatch
- simplifies overview slides and proof-language wording
- focuses the presentation on the actual HeapLang / OCPL proof development,
  removing the custom Jessie-like scaffold from the slide story

dckc committed:
spec: drop custom Jessie scaffold from active tree

Remove the non-HeapLang jessie*.v scaffold files from the active spec tree:
- jessie_lang.v
- jessie_counter.v
- jessie_eval.v
- jessie_robust.v
- jessie_counter_robust.v

Update _CoqProject and iris.mk so the active development is now just:
- the modernized OCPL port under ocpl/
- the direct HeapLang case study in jessie_counter_heaplang.v

Verification:
- ocpl/modern_heap.v, modern_lifting.v, modern_on_val.v, and modern_robust_safety.v still compile
- the remaining build still fails in the existing local edits to jessie_counter_heaplang.v at line 53; this commit does not touch that file

dckc committed:
spec: generalize HeapLang counter proof over monotone attackers

Refactor the direct HeapLang counter proof so the checked-client theorem is no
longer hard-coded to the hole and single-incr attackers.

This commit:
- factors the counter methods as incr_cap / decr_cap helpers
- proves checked_client_from_mono_attacker_wp, a higher-order theorem saying
  any attacker that preserves the counter cell's lower bound when given incr_cap
  also preserves the final checked-client assertion
- re-expresses the hole and single-incr proofs as instances of that theorem
- restores the active HeapLang build to green

This is still not the full OCPL-style all-contexts theorem, but it isolates the
remaining gap cleanly: derive the monotone-attacker premise from the allowed
attacker-context class.

dckc committed:
spec: port modern on-val rules for unary ops and conditionals

Restore the OCPL-style value-elimination lemmas that are needed on the pure side of the modern robust-safety port.

This adds:
- low_un_op_eval
- wp_on_val_un_op and wp_on_val_un_op_bind
- wp_any_if and wp_any_if_bind

I intentionally did not keep a generalized binary-operator rule in this commit. Modern HeapLang has OffsetOp and extra value cases that do not line up cleanly with the old OCPL story, and that proof was turning into a separate design problem. The current commit keeps the library green while restoring the parts that are directly on the theorem path.

Validation:
- make -f iris.mk build

dckc committed:
spec: extend modern robust-safety confined lemmas

Push the OCPL port one layer up from the raw on-val rules by adding more confined-expression lemmas in modern_robust_safety.

Added:
- confined_lit for literal values
- confined_un_op
- confined_if

These sit alongside the existing app/pair/fst/snd/inj lemmas and make the remaining gap more explicit: we still need the mutually recursive FTLR / value story and a clean way to thread branch proofs through case-analysis continuations.

I intentionally left confined_case as a TODO in this commit. The current wp_on_val_case_bind continuation is latered, and the obvious proof shape wants to reuse non-persistent branch WPs across that later. That is the next proof-engineering issue to solve rather than something to paper over.

Validation:
- make -f iris.mk build

dckc committed:
spec: default iris.mk to the OCPL upstream Coq version

Set COQ_VER to 8.9.1 to match the current upstream ocpl-coq README. Also stop assuming that the old-toolchain default should install modern Iris and vsrocq packages: when COQ_VER=8.9.1 the default EXTRA_OPAM_PKGS is empty, while newer Coq versions still pick up the modern coq-iris/coq-iris-heap-lang/vsrocq set. Update the help text so the install target and defaults are explicit.

dckc committed:
spec: isolate OCPL toolchain in a versioned opam switch

Change the default OPAM_SWITCH from the generic name iris to ocpl-coq-, so the old OCPL/Coq 8.9.1 setup does not collide with an existing modern Iris switch. This makes make -f iris.mk switch/install default to a separate old-toolchain environment without requiring manual overrides.

dckc committed:
spec: use an old OCaml compiler for the OCPL 8.9.1 toolchain

The previous old-toolchain default still created an opam switch on ocaml-system, which on this machine meant OCaml 4.14.1 and made coq.8.9.1 unsatisfiable. For the OCPL path, default OCAML_VER to 4.07.1, make the default switch name include both Coq and OCaml versions, and create the switch with ocaml-base-compiler. instead of ocaml-system. Also show COQ_VER, OCAML_VER, and OPAM_SWITCH in make -f iris.mk help so the selected toolchain is obvious.

dckc committed:
spec: record the recovered OCPL checkout provenance

Note in ocpl/README.md that the upstream OCPL development was rebuilt locally under packages/zoe/spec/_research/ocpl-coq, with the user-reported build timestamp 2020-02-02 12:39 and upstream git hash 922c0f4c. This gives the modern port notes a concrete reference point now that the old VM link is dead.

dckc commented:
## formally verified robust-safety makeCounter, I think.

2026-04-20 13:50 c76b7bca tests: add a robust-safety makeCounter counter example
2026-04-20 13:57 6c8a0089 tests: refine the makeCounter robust-safety example

2026-04-20 14:11 0c600a6d tests: model makeCounter with a plain shared count cell
2026-04-20 14:35 22159803 tests: streamline makeCounter closures and update notation

That's in https://codeberg.org/dckc/ocpl-coq (should be agoric-labs or something, I suppose)
it's a clone of https://gitlab.mpi-sws.org/FCS/ocpl-coq , the work from the 2017 paper.

It didn't fit into agoric-sdk right away.

dckc committed:
spec: retell Jessie/Iris deck around recovered OCPL and source-level bridge

dckc committed:
spec: clarify theorem slide with robust-safety reading

dckc commented:
tab dump from yesterday's discussion...

[agoric-sdk/packages/zoe/spec/jessie-iris.md at dc-escrow-formal · Agoric/agoric-sdk](https://github.com/Agoric/agoric-sdk/blob/dc-escrow-formal/packages/zoe/spec/jessie-iris.md)

[dckc/ocpl-coq: mirror of https://gitlab.mpi-sws.org/FCS/ocpl-coq - Codeberg.org](https://codeberg.org/dckc/ocpl-coq)
[ocpl-coq/theories/jessie/make_counter.v at dc-jessie - dckc/ocpl-coq - Codeberg.org](https://codeberg.org/dckc/ocpl-coq/src/branch/dc-jessie/theories/jessie/make_counter.v)
[ocpl-coq/theories/jessie/escrow2013.v at dc-jessie - dckc/ocpl-coq - Codeberg.org](https://codeberg.org/dckc/ocpl-coq/src/branch/dc-jessie/theories/jessie/escrow2013.v)
[Jessie/packages/parse/src/quasi-jessie.js.ts at main · endojs/Jessie](https://github.com/endojs/Jessie/blob/main/packages/parse/src/quasi-jessie.js.ts)

[FCS / ocpl-coq · GitLab](https://gitlab.mpi-sws.org/FCS/ocpl-coq)


[Iris Project](https://iris-project.org/)

[Matthias Felleisen - Wikipedia](https://en.wikipedia.org/wiki/Matthias_Felleisen)
[Publications](https://felleisen.org/matthias/papers.html)

[amutake/actario: Verification Framework for Actor Systems on Coq](https://github.com/amutake/actario)

[Erlang semantics in Coq](https://epsrc-stardust.github.io/assets/slides/glasgow23/thompson.pdf)

----
p.s.

[Use case: ERTP / Chainmail · Issue #13 · rchain-community/behavr](https://github.com/rchain-community/behavr/issues/13)

[HolisticSpecifications/Rocq/parameterized_chainmail_coq at master · sophiaIC/HolisticSpecifications](https://github.com/sophiaIC/HolisticSpecifications/tree/master/Rocq/parameterized_chainmail_coq)

[FP / RobustDimSum · GitLab](https://gitlab.mpi-sws.org/FP/robustdimsum/-/tree/POPL26)
